### PR TITLE
Threads: use new libuv thread calls

### PR DESCRIPTION
**Pull request by kpamnany**
Removed the pthread calls for thread creation and affinity setting and replaced them with libuv thread calls (from https://github.com/JuliaLang/libuv/pull/31).

Note: I had to manually pull libuv and force rebuild with `rm -rf deps/build/libuv`. How do you ask/force the build system to do this automatically?

### CODE REVIEW

**Comment by vtjnash** [comment]
changing `deps/libuv.version` will get the build system to do that for you (also then run `make configure-libuv NO_GIT=1` to compute the hash and commit those too)

**Comment by kpamnany** [comment]
``deps/libuv.version`` has ``LIBUV_BRANCH`` which hasn't changed (still ``julia-uv.0.11.26``), and ``LIBUV_SHA1``. Running ``make configure-libuv NO_GIT=1`` in ``deps`` doesn't seem to recompute the hash. Clarify further please?

**Comment by vtjnash** [comment]
the important variable is SHA1, not BRANCH

**Comment by tkelman** [comment]
If you get it working, be sure to `git rm` the old libuv hashes in addition to `git add`ing the new ones.

**Comment by kpamnany** [comment]
Review please?

**Comment by tkelman** [comment]
Should we also include https://github.com/JuliaLang/libuv/pull/33 ?

**Comment by kpamnany** [comment]
The AppVeyor build failure is from running out of time. The Travis failures are due to https://github.com/JuliaLang/libuv/pull/33; there's a C99-style declaration at ``src/unix/stream.c:807`` (see https://github.com/JuliaLang/libuv/pull/33/files#diff-c9c9805d800eb132a191a9e40da6d813R807) and ``libuv`` uses C89.

**Comment by tkelman** [comment]
https://github.com/JuliaLang/libuv/commit/e305f89a4aadfcd8a69139de9b7261b8882b429d

Looks like one of the test workers stalled early, in one of the linalg tests? Not sure if related.

**Comment by kpamnany** [comment]
The Makefile git-fetch command is not working for some reason; it isn't fetching https://github.com/JuliaLang/libuv/commit/e305f89a4aadfcd8a69139de9b7261b8882b429d.
```
~/julia/deps/srccache/libuv.git$ git fetch -v git://github.com/JuliaLang/libuv.git julia-uv0.11.26:remotes/origin/julia-uv0.11.26
From https://github.com/JuliaLang/libuv
 = [up to date]      julia-uv0.11.26 -> origin/julia-uv0.11.26
~/julia/deps/srccache/libuv.git$ git log --oneline
a113009 Merge pull request #33 from JuliaLang/amitm/2gb
b8cfb46 Merge pull request #31 from kpamnany/thread-affinity
...
```
If I do a simple ``git fetch``, it brings in that commit:
```
~/julia/deps/srccache/libuv.git$ git fetch
From https://github.com/JuliaLang/libuv
   a113009f..e305f89a  julia-uv0.11.26 -> julia-uv0.11.26
```
Why is this happening?

**Comment by kpamnany** [comment]
I noticed that the AppVeyor build worked fine before including https://github.com/JuliaLang/libuv/pull/33 (see the [log](https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.10603) for https://github.com/JuliaLang/julia/commit/82acd125990722b4a20ebc09e3d61e522d07a6dc). Maybe that code needs further testing on Windows?

**Comment by tkelman** [comment]
That PR only touches unix only code? I think I've been seeing test freezes elsewhere. In the latest log here worker 2 got as far as mpfr but then hung.

**Comment by kpamnany** [comment]
You're right. Must be unrelated then.

**Comment by kpamnany** [comment]
Any comments?

**Comment by vtjnash** [comment]
> The Makefile git fetch command with a refspec is not working for some reason

i believe i've found and fixed the cause of this https://github.com/JuliaLang/julia/commit/a7210451843c7718b1ca9b0dc7c81aee24a307f7

**Comment by kpamnany** [comment]
Awesome. Thank you! I think that was the only problem with this PR.

**Comment by kpamnany** [comment]
Okay to merge?

**Comment by vtjnash** [comment]
it appears that there's a compiler warning in libuv that needs to be fixed:
```c
c:/projects/julia/deps/srccache/libuv-e305f89a4aadfcd8a69139de9b7261b8882b429d/src/uv-common.c:310:44: warning: passing argument 3 of 'GetProcessAffinityMask' from incompatible pointer type
   if (!GetProcessAffinityMask(hproc, &pam, &sam))
                                            ^
In file included from c:/projects/julia/mingw64/x86_64-w64-mingw32/include/windows.h:70:0,
                 from c:/projects/julia/mingw64/x86_64-w64-mingw32/include/winsock2.h:23,
                 from c:/projects/julia/deps/srccache/libuv-e305f89a4aadfcd8a69139de9b7261b8882b429d/include/uv-win.h:32,
                 from c:/projects/julia/deps/srccache/libuv-e305f89a4aadfcd8a69139de9b7261b8882b429d/include/uv.h:62,
                 from c:/projects/julia/deps/srccache/libuv-e305f89a4aadfcd8a69139de9b7261b8882b429d/src/uv-common.c:29:
c:/projects/julia/mingw64/x86_64-w64-mingw32/include/winbase.h:1124:29: note: expected 'PDWORD_PTR' but argument is of type 'DWORD *'
   WINBASEAPI WINBOOL WINAPI GetProcessAffinityMask (HANDLE hProcess, PDWORD_PTR lpProcessAffinityMask, PDWORD_PTR lpSystemAffinityMask);
                             ^
```

`PDWORD_PTR` is `*DWORD_PTR` is `*ULONG_PTR` is `*uintptr_t`
the types of pam, sam, tam, and oldtam in `uv_thread_setaffinity` need to changed accordingly.

**Comment by kpamnany** [comment]
So essentially:
```
diff --git i/src/uv-common.c w/src/uv-common.c
index 1436309f..d36bf492 100644
--- i/src/uv-common.c
+++ w/src/uv-common.c
@@ -302,7 +302,7 @@ int uv_thread_setaffinity(uv_thread_t *tid,
 #ifdef _WIN32
   int i;
   HANDLE hproc;
-  DWORD pam, sam, tam = 0, oldtam;
+  DWORD_PTR pam, sam, tam = 0, oldtam;
 
   assert(mask_size >= UV_CPU_SETSIZE);
```
Should I push this directly?

**Comment by vtjnash** [comment]
Yes, lgtm

File: src/threading.c [hunk 1]
```diff
@@ -240,9 +207,9 @@ void jl_init_threading(void)
 
 void jl_start_threads(void)
 {
-    char *cp;
+    char *cp, mask[UV_CPU_SETSIZE];
```
Comment (tkelman) [inline src/threading.c]: I'm getting `error: 'UV_CPU_SETSIZE' undeclared` here

Comment (kpamnany) [inline src/threading.c]: On what platform? It's defined at https://github.com/JuliaLang/libuv/blob/julia-uv0.11.26/include/uv.h#L2257.

Comment (tkelman) [inline src/threading.c]: ah sorry, my mistake - was in a docker container (ubuntu 14.04 x86_64) with a slightly unusual configuration that didn't have libuv rebuilding properly, nevermind

**Pull request merged**